### ISSUE TITLE
If an item is opened in 2 tabs and I do modifications in one of the tab and save, modifications in the second tab will override my first changes

### ISSUE THREAD

**Issue by akramsarkis** [openpim/server#41]
- [ ] If an item is opened in 2 tabs and I do modifications in one of the tab and save, modifications in the second tab will override my first changes

1. Have product 1 opened in  2 tabs
2. In the first tab do modifications and save
3. In the second tab do modifications and save, you will see that the changes done in the first tab are no more there

It will be better to have when saving in the second tab a warning that changes has been done and will be lost

**Comment by jfayad** [comment]
this is actually having an impact on what I'm doing in #38 

I noticed it with the following scenario
1. Start a DB transaction and put a lock on an item
2. Keep the lock on and open the item through the UI
3. Make a change and save the item in the UI (the UI will hang in waiting for the request to respond)
4. Finalize the DB transaction and release the lock
5. The Save operation in the UI proceeds but it overrides the changes made in the DB Transaction

An easy fix here would be to require the updatedAt field as an arg of the item being changed and check its value is the same as the one for the item we have in DB.
Another, simpler and more robust, option could be to explicitly request a lock when editing an item with the NOWAIT option so the request will fail instead of hanging

I honestly can't think of any business situation where allowing users to override the changes of another user/process, without them being fully aware of what they're doing, go right.

**Comment by openpim** [comment]
sorry, but we for sure will not use locks... this is bad approach because it will affected a lot of performance.
If you have thousands of records to upload through Excel for example, or by API it will slow system a lot.
Plus it will require a lot of changes in the core. You will have to lock the record, then free it.
And we have actions that can run during update time, so all this will add a lot of complexity.

I was thinking about optimistic locking approach when second user will receive the error that data is already modified and he will be able to save current data and reload the object again. This is much easier to make.

**Comment by jfayad** [comment]
Actually PostgreSQL is automatically putting a lock on the row when it's being updated.
Requesting a lock explicitly with the NOWAIT option will not change this behavior but it will fail instead of waiting for the lock to be freed.

It can surely introduce other side effects (with more error management) but I personally think these cases will only surface and force the development to explicitly handle the cases instead of leaving it up to luck.

That could be a parameter one can pick when deploying his own instance of the PIM ?

**Comment by openpim** [comment]
yes, but this will not help you in this case. You date data loaded in 2 tabs. Then you press save at one tab. You will block the record in DB, update it and free the lock. Then click save on second tab and data will be overridden the same as now (also lock record, update with the data from second tab, free lock). Because first lock will be already free.
 
You should have some kind of lock button at the screen, so user can lock the data, edit it for some time and then free it. And when second user try to lock the record it received the error. This is completely another case of work. And really I do not like it. What if user forgot to free the record, etc?

**Comment by jfayad** [comment]
@openpim the issue with two tabs opened is what @akramsarkis raised, but I was referring to the point I mentioned in #38 

But indeed, with the bigger picture of both situation you are right, the soft lock based on the UpdatedAt date should work for both cases, this means that the updatedAt file should be provided as a param to the mutation though, and if it's missing the mutation should fail.

I also thing the approach of a button on the UI is a bad one, as you said.
